File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """

  @spec count(String.t) :: map
  def count(sentence) do
    words = String.downcase(sentence)
    |> String.split(~r/[^[:alpha:][:digit:]-]/u, trim: true)
    |> Enum.reduce(%{}, &create_map/2)
  end

  @spec create_map(String.t, map) :: map
  defp create_map(word, map) do
    Map.update(map, word, 1, fn count -> count + 1 end)
  end
end

File: ecmascript/word-count/word-count.js
export default class Words {
  count(string) {
    const words = string.split(/\s/);
    return words.reduce((wordCounts, word) => {
      const trimmed = word.trim().toLowerCase();
      const currentWord = wordCounts[trimmed];
       if (trimmed.length < 1) {
         return wordCounts;
       }

       if (currentWord && typeof currentWord === 'number') {
         wordCounts[trimmed] += 1;
       } else {
        wordCounts[trimmed] = 1;
       }

       return wordCounts;
    }, {});
  }
}